File: Platonov/1376.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        int* degree = new int[n];
        int* maxTimeToInform = new int[n];
        for (int i = 0; i < n; ++i) {
            degree[i] = 0;
            maxTimeToInform[i] = 0;
        }
        for (int i = 0; i < n; ++i) {
            if (manager[i] != -1) {
                ++degree[manager[i]];
            }
        }
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 0) {
                int v = i;
                while (manager[v] != -1) {
                    maxTimeToInform[manager[v]] = max(maxTimeToInform[manager[v]], informTime[manager[v]] + maxTimeToInform[v]);
                    if (degree[manager[v]] == 1) {
                        v = manager[v];
                    } else {
                        --degree[manager[v]];
                        break;
                    }
                }
            }
        }
        int result = 0;
        for (int i = 0; i < n; ++i) {
            result = max(result, maxTimeToInform[i]);
        }
        return result;
    }
};

File: Nenakhov/1376.py
class Solution(object):
    def numOfMinutes(self, n, headID, manager, informTime):
        """
        :type n: int
        :type headID: int
        :type manager: List[int]
        :type informTime: List[int]
        :rtype: int
        """
        descendants = [set({}) for i in range(n)]
        BFS_queue = collections.deque()
        employee_time = [0 for i in range(n)]

        BFS_queue.append(headID)
        employee_time[headID] = 0

        for i in range(n):
            if(manager[i]!=-1):
                descendants[manager[i]].add(i)

        while(BFS_queue):
            employee = BFS_queue.popleft()

            for i in descendants[employee]:
                BFS_queue.append(i)
                employee_time[i] = employee_time[employee] + informTime[employee]
        
        return max(employee_time)
